libs/json/encoder_test.go

```diff
@@ -0,0 +1,99 @@
+		"uint64":          {uint64(64), `"64"`},
+		"time":            {ti, `"2020-06-02T16:05:13.004346374Z"`},
+		"time empty":      {time.Time{}, `"0001-01-01T00:00:00Z"`},
+		"time ptr":        {&ti, `"2020-06-02T16:05:13.004346374Z"`},
+		"customptr":       {CustomPtr{Value: "x"}, `{"Value":"x"}`}, // same as encoding/json
+		"customptr ptr":   {&CustomPtr{Value: "x"}, `"custom"`},
+		"customvalue":     {CustomValue{Value: "x"}, `"custom"`},
+		"customvalue ptr": {&CustomValue{Value: "x"}, `"custom"`},
+		"slice nil":       {[]int(nil), `null`},
+		"slice bytes":     {[]byte{1, 2, 3}, `"AQID"`},
+		"slice int64":     {[]int64{1, 2, 3}, `["1","2","3"]`},
+		"slice int64 ptr": {[]*int64{&i64, nil}, `["64",null]`},
+		"array int64":     {[3]int64{1, 2, 3}, `["1","2","3"]`},
+		"map int64":       {map[string]int64{"a": 1, "b": 2, "c": 3}, `{"a":"1","b":"2","c":"3"}`},
+		"car":             {car, `{"type":"vehicle/car","value":{"Wheels":4}}`},
+		"car value":       {*car, `{"type":"vehicle/car","value":{"Wheels":4}}`},
+		"car iface":       {Vehicle(car), `{"type":"vehicle/car","value":{"Wheels":4}}`},
+		"boat":            {boat, `{"type":"vehicle/boat","value":{"Sail":true}}`},
+		"boat ptr":        {&boat, `{"type":"vehicle/boat","value":{"Sail":true}}`},
+		"boat iface":      {Vehicle(boat), `{"type":"vehicle/boat","value":{"Sail":true}}`},
+		"key public":      {PublicKey("12345678"), `{"type":"key/public","value":"MTIzNDU2Nzg="}`},
+		"tags": {
+			Tags{JSONName: "name", OmitEmpty: "foo", Hidden: "bar", Tags: &Tags{JSONName: "child"}},
+			`{"name":"name","OmitEmpty":"foo","tags":{"name":"child"}}`,
+		},
+		"tags empty": {Tags{}, `{"name":""}`},
+		// The encoding of the Car and Boat fields do not have type wrappers, even though they get
+		// type wrappers when encoded directly (see "car" and "boat" tests). This is to retain the
+		// same behavior as Amino. If the field was a Vehicle interface instead, it would get
+		// type wrappers, as seen in the Vehicles field.
+		"struct": {
+			Struct{
+				Bool: true, Float64: 3.14, Int32: 32, Int64: 64, Int64Ptr: &i64,
+				String: "foo", StringPtrPtr: &sPtr, Bytes: []byte{1, 2, 3},
+				Time: ti, Car: car, Boat: boat, Vehicles: []Vehicle{car, boat},
+				Child: &Struct{Bool: false, String: "child"}, private: "private",
+			},
+			`{
+				"Bool":true, "Float64":3.14, "Int32":32, "Int64":"64", "Int64Ptr":"64",
+				"String":"foo", "StringPtrPtr": "string", "Bytes":"AQID",
+				"Time":"2020-06-02T16:05:13.004346374Z",
+				"Car":{"Wheels":4},
+				"Boat":{"Sail":true},
+				"Vehicles":[
+					{"type":"vehicle/car","value":{"Wheels":4}},
+					{"type":"vehicle/boat","value":{"Sail":true}}
+				],
+				"Child":{
+					"Bool":false, "Float64":0, "Int32":0, "Int64":"0", "Int64Ptr":null,
+					"String":"child", "StringPtrPtr":null, "Bytes":null,
+					"Time":"0001-01-01T00:00:00Z",
+					"Car":null, "Boat":{"Sail":false}, "Vehicles":null, "Child":null
+				}
+			}`,
+		},
+	}
```

Comment:
This really small and quickly written package is at least as well tested as amino if not even better 👍 

Regarding documentation, I have a suggestion: Maybe a very exhaustive [example style test](https://blog.golang.org/examples) inside a `doc.go` would be the best documentation?
It's probably faster to write than well-written and well-structured english documentation and is to the point for everyone looking for answers about edge-cases in documentation.

Reply:
Good idea.

Reply:
Ok, so I tried the example test approach, but it's a bit tricky since it only allows a _single_ output for a test. So this won't work, because only the last `Output` takes effect:

```go
func Example() {
	printJSON("foo")     // Output: "foo"
	printJSON(int64(64)) // Output: "64"
	printJSON(int32(32)) // Output: 32
}
```

```
=== RUN   Example
--- FAIL: Example (0.00s)
got:
"foo"
"64"
32
want:
32
FAIL
FAIL	github.com/tendermint/tendermint/libs/json	0.241s
FAIL
```

I can add one example test per case:

```
func ExampleMarshal_i32()    { printJSON(int32(32)) /* Output: 32 */ }
func ExampleMarshal_i64()    { printJSON(int64(64)) /* Output: "64" */ }
func ExampleMarshal_string() { printJSON("foo") /* Output: "foo" */ }
```

But this results in very verbose documentation, with 4 paragraphs of doc output for every single test.

I'll just add a doc comment instead.

Reply:
I see 👍 

An alternative approach could be to group a bunch of basic types into an array and (pretty)print them into a single output. 

The edge cases (`null`, `Time` etc) could deserve separate paragraphs even. But a doc comment is probably as good as this. I only suggested  Example tests because I thought they'll be faster to write actually (yet contain all relevant info).

Reply:
I've added package-level documentation, let me know if you think anything is lacking. Will merge after lunch.

Reply:
LGTM 👍